File: code/107.binary-tree-level-order-traversal-ii.py
#
# @lc app=leetcode id=107 lang=python3
#
# [107] Binary Tree Level Order Traversal II
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def traverse(self, root, level):
        if root == None:
            return
        if level == len(self.list):
            self.list.insert(0, [])
        self.list[-(level+1)].append(root.val)
        self.traverse(root.left, level+1)
        self.traverse(root.right, level+1)

    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        self.list = []
        self.traverse(root, 0)
        return self.list

        

# @lc code=end

File: code/107.binary-tree-level-order-traversal-ii.cpp
/*
 * @lc app=leetcode id=107 lang=cpp
 *
 * [107] Binary Tree Level Order Traversal II
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void helper(TreeNode* node, int level, vector<vector<int>> &theVector) {
        if(!node) return;
        int size = theVector.size();
        ++level;
        if(level > size) {
            theVector.push_back(vector<int>());
        }
        helper(node->left, level, theVector);
        helper(node->right, level, theVector);
        theVector[level - 1].push_back(node->val);
    }

    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> theVector;
        helper(root, 0, theVector);
        reverse(theVector.begin(), theVector.end());
        return theVector;
    }
};

